pulser/waveforms.py

```diff
@@ -538,13 +538,27 @@ def from_max_val(
             raise ValueError(
                 "The maximum value and the area must have " "matching signs."
             )
+
         # A normalized Blackman waveform has an area of 0.42 * duration
         duration = np.ceil(area / (0.42 * max_val) * 1e3)  # in ns
         wf = cls(duration, area)
+
         # Adjust for rounding errors to make sure max_val is not surpassed
         while np.abs(wf._scaling) > np.abs(max_val):
             duration += 1
+            previous_wf = wf
             wf = cls(duration, area)
+
+        # According to the documentation for numpy.blackman(), "the value one
+        # appears only if the number of samples is odd". Hence, the previous
+        # even duration can reach a maximal value closer to max_val.
+        if (
+            previous_wf is not None
+            and duration % 2 == 1
+            and np.max(wf.samples) < np.max(previous_wf.samples) <= max_val
```

Comment:
This check might be the most costly and it will fail when `previous_wf = None`, so you should nest it as a second `if`-statement.

Reply:
According to https://docs.python.org/3.8/reference/expressions.html#boolean-operations:

`The expression 'x and y' first evaluates x; if x is false, its value is returned; otherwise, y is evaluated and the resulting value is returned.`

Hence, no further condition is tested as soon as one fails. That's why I ordered them in increasing computation costs.